### PR TITLE
DISP-3652: Add search endpoint to dispatch

### PR DESCRIPTION
**Pull request by derekcaneja**

### CODE REVIEW

File: lib/dispatch.js [hunk 1]
```diff
@@ -233,4 +235,138 @@ export default class Dispatch {
...
+  /**
+   * Search an organization's entities.
+   * @param  {[String]}  options.organizationID The organization's id.
+   * @param  {[String]}  options.query          The search text.
+   * @param  {[String]}  options.filter         The entity to filter by.
+   * @param  {[Boolean]} options.skipCustomers  Whether or not to show the customers
+   * @param  {[Number]}  options.maxResults     The maximum results to show.
+   * @return {Promise}                          A promise with the results.
+   */
+  search({ organizationID, query, filter, skipCustomers, maxResults }) {
```
Comment (aldeed) [inline lib/dispatch.js]: I didn't realize when we were pairing, but actually `filter` now removes the need for `skipCustomers`. To skip searching customers, you would do `filter: 'Job,JobOffer'`

And on that subject, it looks like the API expects a comma list like `models=Job,JobOffer`, but passing in `filter: 'Job,JobOffer'` is a little strange, so maybe `filter` can be an array, and you can add some code here to do `filter.join(',')`

File: lib/dispatch.js [hunk 2]
```diff
@@ -233,4 +235,138 @@ export default class Dispatch {
...
     return url;
   }
+
+  /**
+   * Search an organization's entities.
+   * @param  {[String]}  options.organizationID The organization's id.
+   * @param  {[String]}  options.query          The search text.
+   * @param  {[String]}  options.filter         The entity to filter by.
+   * @param  {[Boolean]} options.skipCustomers  Whether or not to show the customers
+   * @param  {[Number]}  options.maxResults     The maximum results to show.
```
Comment (aldeed) [inline lib/dispatch.js]: "The maximum results to include per model"

Should we call it `maxResultsPerModel` for clarity?

File: lib/dispatch.js [hunk 3]
```diff
@@ -233,4 +235,138 @@ export default class Dispatch {
...
+          totalResults = jobIds.length;
+        } else if (filter === 'Job Offers') {
+          totalResults = jobOfferIds.length;
+        } else if (filter === 'Customers') {
+          totalResults = customers.length;
+        } else {
+          customers.length = capResultsLength(customers.length);
+          jobIds.length = capResultsLength(jobIds.length);
+          jobOfferIds.length = capResultsLength(jobOfferIds.length);
+        }
```
Comment (aldeed) [inline lib/dispatch.js]: This whole `if` and `capResultsLength` function can be removed. The API should be limiting the results now and respecting the filters we request so `totalResults = totalCustomers + totalJobs + totalJobOffers;` should already give us the correct answer.

File: lib/dispatch.js (hunk 2 above)
Comment (derekcaneja) [inline lib/dispatch.js]: Works for me.

File: lib/dispatch.js [hunk 4]
```diff
@@ -233,4 +235,138 @@ export default class Dispatch {
...
+        let message;
+        if (err.errors) {
+          message = err.errors[0] && err.errors[0].message;
+        } else if (err.message) {
+          message = err.message.toString();
+        } else {
+          message = 'Unexpected error';
+        }
+        throw new Error(message);
+      });
```
Comment (aldeed) [inline lib/dispatch.js]: It looks like the `.get` calls will already clean up and throw any API errors, so I would remove this whole `catch`.

https://github.com/DispatchMe/dispatch-node-sdk/blob/master/lib/rawClient.js#L132-L145

**Comment by aldeed** [comment]
Left some comments

**Comment by aldeed** [comment]
👍

**Pull request merged**